### PR TITLE
Correct ARM STM32 I2C frequency.

### PR DESCRIPTION
**Pull request by yiancar**
It was beleaved that this setting result in a 400Khz I2C bus.

This was incorrect, actual frequency measure with a logic analyzer was around 150Khz.

This is derived from the excel sheet linked in the .h file.
Also confirmed with the ST IDE.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: drivers/arm/i2c_master.h [hunk 1]
```diff
@@ -73,19 +73,19 @@
 // The default timing values below configures the I2C clock to 400khz assuming a 72Mhz clock
```
Comment (yanfali) [inline drivers/arm/i2c_master.h]: This comment needs fixing

**Comment by yiancar** [comment]
Fixed.

File: drivers/arm/i2c_master.h (hunk 1 above)
Comment (yiancar) [reply drivers/arm/i2c_master.h]: No this is correct:)
The setting were giving 150khz we wanted 400

Comment (yanfali) [reply drivers/arm/i2c_master.h]: My only concern with this change is it may break splits. We've had issues in the past with splits running at 400KHz failing because of excessively long trrs cables.

**Comment by yanfali** [comment]
@zvecr you are working in this area, can you weigh in?

**Comment by yiancar** [comment]
yes..:)

**Comment by ashandore** [comment]
If the CPU clock speed is available at compile time (looks like the `F_CPU` macro?), this could calculate register values to produce desired i2c clock rate of 400kHz given the firmware's actual clock configuration. This would be more robust to changes in core frequency, and as long as the calculation only uses literals & literal macros the i2c configuration object can remain `const`.

Calculations are available in part reference manuals and maybe from the referenced spreadsheet (I don't have access to office apps at the moment), for example:

image

where tI2CCLK is the i2c peripheral clock period. It appears that the code currently assumes that I2CCLK is derived directly from F_CPU with no prescaling; it could continue to do so if that's a reasonable assumption.

**Comment by zvecr** [comment]
Calculating these values would be my end goal preference (or at worst case providing tables like we do for stuff like AVR backlight). The current setup makes it difficult to work correctly across supported MCU as it really targets a f303, which will become less ideal with time. I have a ticket for ARM split where we want to run at 400k normally and downgrade to 100k when using split. The current solution makes that more difficult for I2Cv2.

Outside that, I dont currently have access to Excel to validate the new values, and Google Docs/Libreoffice wont open the ST provided file.

**Comment by yiancar** [comment]
The problem is is that those values change per mcu, as for example in mcuconfig.h the clock tree variables are set which setup the end clock frequency, but it all depends on the clock you use etc. 

TL;DR F_CLK is not available but rather the dividers of the clock tree. (this is because of the vast number of MCUs supported.

Also giving people tables or calculations is a bit dangerous as someone would like to setup hold and setup timing depending on their application. @zvecr My approach on this is: Download the ST IDE (0.5GB) and it does all the clock tree, i2c or whatever other magic math you want for you.

The above is also I think the best approach with someone developing something with CHIBIOS and STM32 due to the wild number of targets.

TL;DR there is so many mcu targets and so many configurations of clock trees, peripherals etc that getting the ST IDE to give you the magic values is not only the industry standard but also makes lots of sence

**Comment by zvecr** [comment]
I will have a look at your ST IDE suggestion.

As for the topic of computing values, its about providing sane defaults. Not everyone wants or should need to be a ChibiOS expert. If we can provide a good user XP to devs, while also allowing overrides like the existing ones for advanced users its the best of both worlds.

**Comment by yiancar** [comment]
I understand, the problem is there need to be assumptions:
0) is this an F303?
1) did you use an 8Mhz crystal?
2) did you set the clock tree to result in a 72Mhz master clock?
3) do you want 0 setup delay and 0 hold delay?
4) did you setup the clock tree to use master clock as i2c input clock?

So sadly it is not as simple as use 400KHz if you think about it.

**Comment by ashandore** [comment]
It's typical for vendor HALs to provide a means for getting runtime information about peripheral clocking (and this is certainly true for ST's HAL). Normally my recommendation these APIs to calculate timing parameters at runtime based on user input like "I want 400kHz". I am unfamiliar with ChibiOS, though, I do not know if it makes them available or if it would be appropriate to use ST's APIs directly.

If it's desirable for other parameters to be adjustable by users, providing an API that exposes them or requiring more #defines to configure them would be a way to do so.

My `FCLK` suggestion was based on a quick glance at board makefiles; as far as I could tell it's standard for them to define this macro and it appears to be used in other driver code. It certainly doesn't cover all of the assumptions, but if I'm not completely misunderstanding the semantics of `FCLK` it's at least a reduction.

**Comment by yiancar** [comment]
Correct, I think STM32_HCLK can be used to get master clock, then you would need to check how is the mux for the i2c clock is set and you are set. 

I will not be going to integrate this as to be honest in my way of thinking its a bit too much abstraction, but I am willing to help anyone one who leads the initiative.

**Comment by zvecr** [comment]
I think the discussion was about wishes rather than changes for this PR, and for this case, i have added #7096 for us to continue.

**Comment by drashna** [comment]
@zvecr should this be merged in the meanwhile, or do we prefer waiting on #7096?

**Comment by yiancar** [comment]
Hello?:)

**Comment by drashna** [comment]
Honestly, I'll defer to others here as they have more experience with this. 

That said, this doesn't appear to break anything (eg, my planck ez), and it looks good.  

So, really:
`__attribute__((weak)) ✅ `

**Comment by drashna** [comment]
@zvecr I know you have another issue for this, but is there anything that would prevent this from being merged as is?

**Comment by zvecr** [comment]
> when we wrote this we blindly trusted the internet

That's largely what you're asking us to do here :wink:

I see values like `hi2c2.Init.Timing = 0x2000090E;` and  `hi2c2.Init.Timing = 0x0000020B;` within STM32CubeIDE but I'm missing the domain knowledge to translate that to the values you have specified.

**Comment by yiancar** [comment]
So first things first we need the clock tree, this is because you can use either HSI or SYSCLK as I2C clock source. In the default qmk project case it is set to SYSCLK which is set to 72MHz:
clock tree

Then go to the STM32IDE or the magic excel sheet:
Set desired I2C speed to 400KHz, analog filter on (default), no digital filter (default), rise/fall delays to 0ns (we assume that as this is a keyboard the data and clock lines are fairly short):
stm32ide
magic excel

I put both screenshots as one splits the TIMINGR register into our defines.
If you need logic analyzer screenshot for verification let me know and I will get one.

Cheers
